lib/write_full.c

```diff
@@ -0,0 +1,52 @@
+/*
+ * SPDX-FileCopyrightText:  2023, Christian Göttsche <cgzones@googlemail.com>
+ *
+ * SPDX-License-Identifier:  BSD-3-Clause
+ */
+
+
+#include <config.h>
+
+#ident "$Id$"
+
+#include "prototypes.h"
+
+#include <errno.h>
+#include <unistd.h>
+
+
+/*
+ * write_full - write entire buffer
+ *
+ * Write up to count bytes from the buffer starting at buf to the
+ * file referred to by the file descriptor fd.
+ * Retry in case of a short write.
+ *
+ * Returns the number of bytes written on success, -1 on error.
+ */
+ssize_t write_full(int fd, const void *buf, size_t count) {
+	ssize_t written = 0;
+
+	while (count > 0) {
+		ssize_t res;
+
+		res = write(fd, buf, count);
+		if (res < 0) {
+			if (errno == EINTR) {
+				continue;
+			}
+
+			return res;
+		}
+
+		if (res == 0) {
+			break;
+		}
+
+		written += res;
+		buf = (const unsigned char*)buf + res;
+		count -= res;
+	}
+
+	return written;
+}
```

Comment:
I have a few improvements over this implementation:

-  written is a redundant variable.  If you keep buf untouched and then operate on a second pointer, you can subtract both at the end and get the written amount cheaper.  The antipattern was seeing two consecutive `+=res`.
-  `res` is a bit confusing.  It took me several seconds to find that it means result.  I suggest using `w`.  IMHO, it is more obvious that it is just a temporary variable to hold the return value of write(2), as single-letter variables tend to be.
-  While I understand that using `const void *` is a more generic interface, maybe we don't need it.  With `const char *` we would have one less cast.

```c
ssize_t
write_full(int fd, const void *buf, size_t size)
{
	const char  *p0, *p;

	p = p0 = buf;

	while (size > 0) {
		ssize_t  w;

		w = write(fd, p, size);
		if (w == -1) {
			if (errno == EINTR)
				continue;

			return w;
		}

		if (w == 0)
			break;

		p += w;
		size -= w;
	}

	return p - p0;
}
```

Reply:
And it looks like you can drop p0 and just return p - buf (or declare the incoming variable as const char *p0).

Reply:
Yeah, it was a glitch from a previous implementation I had.  :)